include/dlaf/eigensolver/get_tridiag_rank1_nworkers.h

```diff
@@ -0,0 +1,30 @@
+//
+// Distributed Linear Algebra with Future (DLAF)
+//
+// Copyright (c) 2018-2023, ETH Zurich
+// All rights reserved.
+//
+// Please, refer to the LICENSE file in the root directory.
+// SPDX-License-Identifier: BSD-3-Clause
+//
+#pragma once
+
+#include <cstdint>
+
+#include <pika/runtime.hpp>
+
+#include "dlaf/tune.h"
+
+namespace dlaf::eigensolver::internal {
+
+inline std::size_t getTridiagRank1NWorkers() noexcept {
+  const std::size_t nworkers = getTuneParameters().tridiag_rank1_nworkers;
+
+  // Note: precautionarily we leave at least 1 thread "free" to do other stuff
+  const std::size_t max_workers = pika::resource::get_thread_pool("default").get_os_thread_count() - 1;
+
+  // 1 <= number of workers < max_workers
+  return std::max<std::size_t>(1, std::min<std::size_t>(max_workers, nworkers));
```

Comment:
`std::min/max` also need `#include <algorithm>`.

Reply:
Actually the comment above the return is wrong in many ways, but the most important wrong bit is that number of workers is not ensured to be inside that range.

With `max_workers = 0`, that we get when there is a single thread inside the default pool, current implementation does `max(1, min(0, x)) = 1` (given `x >= 0`).

The first trivial `clamp` replacement would look like

```cpp
std::clamp<std::size_t>(nworkers, 1, max_workers);
```

which specifically ends up in the corner case with following values

```cpp
std::clamp<std::size_t>(nworkers, 1, 0); // corner case
```

But from `clamp` doc

> The behavior is undefined if the value of lo is greater than hi.

so we should manage differently this corner case. The solution might be

```cpp
  // `max_workers` should be at least 1
  const std::size_t max_workers = std::max(1, pika::resource::get_thread_pool("default").get_os_thread_count() - 1);
  return std::clamp<std::size_t>(nworkers, 1, max_workers);
```

Anyhow, whatever option we will take, a probably worth discussion to have is if it is a safety requirement to leave 1 rank free from bulk operation or it is a strong requirement (and we require at least two threads in default pool).

Reply:
Good point about the undefined behaviour. In that case the `min`/`max` version that you already had there is possibly the clearest. I'm happy with either version.

Regarding number of threads, I don't think it should be a strong requirement that the default pool has at least two worker threads. If it is, we should make sure to lift that requirement (since it implies we're doing something blocking).

Independent of `clamp` vs. `min`/`max`, couldn't you leave out at least some of the `std::size_t` template parameters:
```suggestion
  // Note: precautionarily we leave at least 1 thread "free" to do other stuff
  const std::size_t max_workers = pika::resource::get_thread_pool("default").get_os_thread_count() - 1;

  // 1 <= number of workers < max_workers
  return std::max(std::size_t(1), std::min(max_workers, nworkers));
```
?

Reply:
what about https://github.com/eth-cscs/DLA-Future/pull/860/commits/8eab3d1ed9b860911810d55ea7039b49fda06acc?